### PR TITLE
Add shadow jar and workspace controller that starts said jar

### PR DESCRIPTION
**Pull request by MrCreosote**
Tested this setup in the Groups service, replacing the workspace controller there

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -267,10 +294,9 @@ dependencies {
 	)
 	testImplementation 'com.arangodb:arangodb-java-driver:6.7.2'
 	testImplementation 'com.github.zafarkhaja:java-semver:0.9.0'
-	testImplementation 'org.hamcrest:hamcrest-core:1.3'
-	testImplementation 'commons-lang:commons-lang:2.4'
```
Comment (MrCreosote) [inline build.gradle:271]: Turns out this isn't needed

Comment (Xiangs18) [reply build.gradle:271]: commons-lang is not needed?

File: build.gradle [hunk 2]
```diff
@@ -132,6 +139,26 @@ task jacocoTestQuickReport(type: JacocoReport, dependsOn: testQuick) {
...
+	// we don't want to package the auth shadow jar, so no test runtime classpath
+	configurations = [project.configurations.runtimeClasspath]
+
+	enableRelocation true
+	relocationPrefix 'us.kbase.workspace.shadow'
+	
+	mergeServiceFiles()
+	
+	from(BUILD_JAVA_DOC_DIR) { into "$IN_JAR_DOCS_DIR/javadoc" }
+	from(BUILD_OTHER_DOC_DIR) { into IN_JAR_DOCS_DIR }
```
Comment (Xiangs18) [inline build.gradle:159]: Are these two lines just copy files into /server_docs?

Comment (MrCreosote) [reply build.gradle:159]: yep

Comment (Xiangs18) [reply build.gradle:159]: so they are in shadow jar. Why we need to do this? Or is just for documentation.

Comment (MrCreosote) [reply build.gradle:159]: So that if someone runs a test against the shadowjar that touches the /docs endpoint for some reason things work as expected

Comment (Xiangs18) [reply build.gradle:159]: how are /docs and /server_docs connected? It somehow knows to look into /server_docs when a test touches /docs endpoint?

Comment (Xiangs18) [reply build.gradle:159]: 👍

**Review by Xiangs18 — APPROVED** [review]
LGTM!
@ialarmedalien should also take a look!

File: build.gradle [hunk 3]
```diff
@@ -15,7 +15,14 @@ plugins {
 	id 'com.github.johnrengelman.shadow' version '8.1.1'
 }
 
-var buildDocsDir = "$buildDir/docs/otherdoc/"
+var BUILD_DOC_ROOT = "$buildDir/docs/"
+var BUILD_JAVA_DOC_DIR = "$BUILD_DOC_ROOT/javadoc"
+var BUILD_OTHER_DOC_DIR = "$BUILD_DOC_ROOT/otherdoc/"
+// This is where the DocServer will look for docs, if it changes it needs to change
+// in deloy.cfg as well
```
Comment (briehl) [inline build.gradle:22]: typo: `deploy.cfg`

File: src/us/kbase/workspace/test/controllers/workspace/WorkspaceController.java [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+		final ProcessBuilder servpb = new ProcessBuilder(command)
+				.redirectErrorStream(true)
+				.redirectOutput(tempDir.resolve("workspace.log").toFile());
+
+		final Map<String, String> env = servpb.environment();
+		env.put("KB_DEPLOYMENT_CONFIG", deployCfg.toString());
+
+		workspace = servpb.start();
+		// TODO TEST add periodic check w/ exponential backoff
+		Thread.sleep(5000);
```
Comment (briehl) [inline src/us/kbase/workspace/test/controllers/workspace/WorkspaceController.java:71]: Is this just to make sure that connections get made properly? Are there asynchronous things (like  `servpb.start()`) that go off in this time that are more of a pain to monitor?

**Review by briehl — APPROVED** [review]
Gotta be honest, I'm not entirely sure how to review here. I looked at the shadow jar docs (https://imperceptiblethoughts.com/shadow/) and the gradle stuff looks fine to me.

So here's the summary if I'm reading it right.

`build.gradle`'s updated to make a shadow jar - a single jar containing the whole workspace service + docs + the new `WorkspaceController`.

`WorkspaceController` can then be invoked to stand up a temporary Workspace service to... do stuff? Does it run its own tests? Does that shadow jar get included with other Java-based services to run a temporary workspace a little more seamlessly than running, say, a Docker image?

Anyway, that looks like (from some other test module) you can just go:
```
final WorkspaceController ws = new WorkspaceController(params);
... do test stuff ...
ws.destroy();
```
which is kinda awesome.

Anyway, if that's the intent, this all seems to hit all the buttons and link together. Hitting approve, but there's a typo and a question there, at your leisure.

**Comment by MrCreosote** [comment]
WorkspaceController can then be invoked to stand up a temporary Workspace service to... do stuff? Does it run its own tests? Does that shadow jar get included with other Java-based services to run a temporary workspace a little more seamlessly than running, say, a Docker image?

It's for cases like this: https://github.com/kbase/groups/blob/master/src/us/kbase/test/groups/integration/ServiceIntegrationTest.java#L124-L132

With the shadowjar, you no longer need to list all the dependent jars: https://github.com/kbase/groups/blob/master/src/us/kbase/test/groups/controllers/workspace/wsjars

So you just add that one jar to your classpath and can easily start up a workspace to test against. 

Eventually I want the workspace repo to publish the shadow jar via jitpack but that's a bit farther down the road

File: src/us/kbase/workspace/test/controllers/workspace/WorkspaceController.java (hunk 4 above)
Comment (MrCreosote) [reply src/us/kbase/workspace/test/controllers/workspace/WorkspaceController.java:71]: The workspace takes a second or two to start up after `ProcessBuilder.start()` is called to check all the connections to other services. This just waits until it should be ready to accept connections.

**Comment by MrCreosote** [comment]
> which is kinda awesome.

and it's **totally** awesome, thank you very much

File: src/us/kbase/workspace/test/controllers/workspace/WorkspaceController.java (hunk 4 above)
Comment (MrCreosote) [reply src/us/kbase/workspace/test/controllers/workspace/WorkspaceController.java:71]: I could do a microoptimization by calling workspace.ver() in a loop waiting for it to come up as per the TODO but it doesn't really seem worth if for now

File: build.gradle (hunk 3 above)
Comment (MrCreosote) [reply build.gradle]: Fixed

**Pull request merged**